### DISCUSSION TITLE
Hyperparameters to tune in coxnet to improve C-index in test dataset

### DISCUSSION THREAD

**Discussion by mldiscussion** [sebp/scikit-survival#254] [Q&A]
We are using Bayesian optimization (hyperopt package) to tune hyperparameters to improve the Coxnet performance.  Currently we only tuned alpha_min_ratio and l1_ratio.  However, the model performance is not good for a dataset we are working on.  Any other parameters that should also tuned and what is the adequate range?  It looks like other hyperparameters that could be tuned include n_alphas.  alphas and penalty_factor do not look like can be tuned, but rather an array should be supplied.  Currently we set them as none.  Thus, it looks like alpha_min_ratio and l1_ratio are two key hyperparameters to be tuned?  The following is the search space we currently used.  Thanks for input!

```
space = {
        'alpha_min_ratio':hp.uniform('alpha_min_ratio', 0.01, 0.99),
        'l1_ratio':hp.uniform('l1_ratio', 0.01, 0.99),
    }

```

**Comment by sebp** [comment]
It is important to keep in mind that CoxnetSurvivalAnalysis fits several models for different penalty factors simultaneously, which are determined automatically via the `alpha_min_ratio` value. Those values will be stored in `fitted_model.alphas_`. During prediction you need to pass it a specific penalty factor (`alpha`), or it will use the last value in `fitted_model.alphas_`, which corresponds to the smallest penalty. Thus, it is likely that many features are selected, if no `alpha` is provided, which in turn might lead to poor performance. To select a single `alpha`, you can use cross-validation as described in the [user guide](https://scikit-survival.readthedocs.io/en/stable/user_guide/coxnet.html#Choosing-penalty-strength-\alpha).

I would consider the other hyper-parameters (`l1_ratio`, `alpha_min_ratio`, `n_alphas`) not crucial. Setting `l1_ratio` to something in the range 0.9-0.99 is usually good. `alpha_min_ratio` determines how small the smallest `alpha` is going to be. For most cases the default is fine. Finally, `n_alphas` determines how many values to explore between the minimum and maximum alpha. It doesn't impact performance itself, but how fine-grained your search for the best alpha is.

Finally, make sure you use sklearn's `StandardScaler` to standardize features before fitting a CoxnetSurvivalAnalysis model.

**Reply by mldiscussion** [reply]
Thank you for the reply.  It looks like `CoxnetSurvivalAnalysis` supports fitting a path of coefficients for varying penalization strength alpha without much computational overhead.  One question is what would be the practical advantage in getting 100 different `alpha` values?  If one of our primary goal is to assess the feature importance, how we should choose the `alpha` value?  I assume if `alpha` value is too high, a lot of features will have coefficient of zero values?

**Reply by sebp** [reply]
As you already mentioned, the main advantage is computational efficiency. Instead of fitting 100 models from scratch, we can fit 100 models in order of regularization strength, which reduces the overall number of iterations required to reach convergence for all 100 fits. Usually, you choose the best `alpha` via cross-validation, as demonstrated in the [user guide](https://scikit-survival.readthedocs.io/en/stable/user_guide/coxnet.html#Choosing-penalty-strength-%5Calpha).

>  I assume if alpha value is too high, a lot of features will have coefficient of zero values?

That is correct.

**Reply by mldiscussion** [reply]
In our analysis, we have found that `alpha_min_ratio`, `l1_ratio` and `alpha` all have impacts on C-index, further `alpha` affect the magnitude of coefficient values and subsequently the hazard ratio for a specific feature.  Is there some way to simultaneously optimize all three parameters, or at least optimize `l1_ratio` and `alpha` by grid search and 5-fold cv? Thanks.

**Reply by sebp** [reply]
Any of sklearn's [hyper-parameter optimizers](https://scikit-learn.org/stable/modules/classes.html#hyper-parameter-optimizers) can be used. Their [user guide](https://scikit-learn.org/stable/modules/grid_search.html#grid-search) contains examples for each of them.
